src/chainbin/nnet3-chain-copy-egs.cc

```diff
@@ -34,8 +36,200 @@ int32 GetCount(double expected_count) {
+                     << "too late to do frame selection or reduce context.";
+          return false;
+        }
+      }
+    }
+  }
+
+  for (int32 i = 0; i < num_indexes_output; i++) {
+    const NnetChainSupervision &outputs = eg.outputs[i];
+    std::vector<Index>::const_iterator iter = outputs.indexes.begin(),
+                                        end = outputs.indexes.end();
+    // Should not have an empty input/output type.
+    KALDI_ASSERT(!outputs.indexes.empty());
+    if (outputs.name == "output") {
+      int32 min_t = iter->t, max_t = iter->t;
+      for (; iter != end; ++iter) {
+        int32 this_t = iter->t;
+        min_t = std::min(min_t, this_t);
+        max_t = std::max(max_t, this_t);
+        if (iter->n != 0) {
+          KALDI_WARN << "Example does not contain just a single example; "
+                     << "too late to do frame selection or reduce context.";
+          return false;
+        }
+      }
+      done_output = true;
+      *min_output_t = min_t;
+      *max_output_t = max_t;
+    } else {
+      for (; iter != end; ++iter) {
+        if (iter->n != 0) {
+          KALDI_WARN << "Example does not contain just a single example; "
+                     << "too late to do frame selection or reduce context.";
+          return false;
+        }
+      }
+    }
+  }
+  if (!done_input) {
+    KALDI_WARN << "Example does not have any input named 'input'";
+    return false;
+  }
+  if (!done_output) {
+    KALDI_WARN << "Example does not have any output named 'output'";
+    return false;
+  }
+  return true;
+}
+
+// calculate the frame_subsampling_factor
+void CalculateFrameSubsamplingFactor(const NnetChainExample &eg,
+                                     int32 *frame_subsampling_factor) {
+  *frame_subsampling_factor = eg.outputs[0].indexes[1].t
+                              - eg.outputs[0].indexes[0].t;
 }
 
+void ModifyChainExampleContext(const NnetChainExample &eg,
+                               int32 left_context,
+                               int32 right_context,
+                               const int32 frame_subsampling_factor,
+                               NnetChainExample *eg_out) {
+  int32 min_input_t, max_input_t,
+        min_output_t, max_output_t;
+  if (!ContainsSingleExample(eg, &min_input_t, &max_input_t,
+                             &min_output_t, &max_output_t))
+    KALDI_ERR << "Too late to perform frame selection/context reduction on "
+              << "these examples (already merged?)";
+  if (left_context != -1) {
+    if (min_input_t > min_output_t - left_context)
+      KALDI_ERR << "You requested --left-context=" << left_context
+                << ", but example only has left-context of "
+                <<  (min_output_t - min_input_t);
+    min_input_t = std::max(min_input_t, min_output_t - left_context);
+  }
+  if (right_context != -1) {
+    if (max_input_t < max_output_t + right_context + frame_subsampling_factor - 1)
```

Comment:
@danpovey  The asymmetric process for right and left is based on the assumption: time of `output `label begin from '0', like (`input_t[0,149](num 150),output_t[0,147](num 50)`),  if the begimning time of output label is not '0', I believe this will crash